### ISSUE TITLE
New rule: Disallow `expression1?.let { expression2 } ?: run { expression3 }` when `expression2` is nullable

### ISSUE THREAD

**Issue by atulgpt** [detekt/detekt#6101]
## Expected Behavior of the rule
The rule should disallow `expression1?.let { expression2 } ?: run { expression3 }` like code as `run` block can be executed either when `expression1` is null or `expression2` returns `null` which is accidental most of the time. Most of new comers replace if else with `null` check with the above expression and they face this issue very later in the cycle

Some example codes as below
```
fun main() {
    val a: String? = "Non null string"
    a?.let {
        println("a is not null case")
        nullable()
    } ?: run {
        println("a is null case")
    }
}

fun nullable(): String? {
    return null
}
```
above code([playground](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS44LjIxIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiLyoqXG4gKiBZb3UgY2FuIGVkaXQsIHJ1biwgYW5kIHNoYXJlIHRoaXMgY29kZS5cbiAqIHBsYXkua290bGlubGFuZy5vcmdcbiAqL1xuXG5mdW4gbWFpbigpIHtcbiAgICB2YWwgYTogU3RyaW5nPyA9IFwiTm9uIG51bGwgc3RyaW5nXCJcbiAgICBhPy5sZXQge1xuICAgICAgICBwcmludGxuKFwiYSBpcyBub3QgbnVsbCBjYXNlXCIpXG4gICAgICAgIG51bGxhYmxlKClcbiAgICB9ID86IHJ1biB7XG4gICAgICAgIHByaW50bG4oXCJhIGlzIG51bGwgY2FzZVwiKVxuICAgIH1cbn1cblxuZnVuIG51bGxhYmxlKCk6IFN0cmluZz8ge1xuICAgIHJldHVybiBudWxsXG59In0=)) prints both `a is not null case` and `a is null case` and hence is problematic
Compliant code should be as below([playground](https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS44LjIxIiwicGxhdGZvcm0iOiJqYXZhIiwiYXJncyI6IiIsIm5vbmVNYXJrZXJzIjp0cnVlLCJ0aGVtZSI6ImlkZWEiLCJjb2RlIjoiLyoqXG4gKiBZb3UgY2FuIGVkaXQsIHJ1biwgYW5kIHNoYXJlIHRoaXMgY29kZS5cbiAqIHBsYXkua290bGlubGFuZy5vcmdcbiAqL1xuXG5mdW4gbWFpbigpIHtcbiAgICB2YWwgYTogU3RyaW5nPyA9IFwiTm9uIG51bGwgc3RyaW5nXCJcbiAgICBpZiAoYSAhPSBudWxsKSB7XG4gICAgICAgIHByaW50bG4oXCJhIGlzIG5vdCBudWxsIGNhc2VcIilcbiAgICAgICAgbnVsbGFibGUoKVxuICAgIH0gZWxzZSB7XG4gICAgICAgIHByaW50bG4oXCJhIGlzIG51bGwgY2FzZVwiKVxuICAgIH1cbn1cblxuZnVuIG51bGxhYmxlKCk6IFN0cmluZz8ge1xuICAgIHJldHVybiBudWxsXG59In0=)) if simple `null` check is intended
```
fun main() {
    val a: String? = "Non null string"
    if (a != null) {
        println("a is not null case")
        nullable()
    } else {
        println("a is null case")
    }
}

fun nullable(): String? {
    return null
}
```

**Comment by atulgpt** [comment]
Hi @cortinico, @BraisGabin can we pick this in detekt 2.0.0? I find recurring issues due to this type of code snippet. I can give this a try

**Comment by BraisGabin** [comment]
I understand the idea but I feel it is way too specific. Don't get me wrong. I love the rules that discourage bad patterns with scope functions. But we should find a more general pattern. For example, is the `run` important here? Or just with an elvis operator is enough? 

This also seems to step a bit on unnecessaryLet. One of my favorite rules. But it is really difficult to define clear bounderies in this type of rules.

**Comment by atulgpt** [comment]
Hi, @BraisGabin for your query,
1. ` is the run important here?` Yes even in cases when `run` is important(for example multi lines expressions) this issue will prevail. I have just taken one line expression to make the example shorter.
```kotlin
fun main() {
    val a: String? = "Non null string"
    a?.let {
        println("a is not null case")
        nullable()
    } ?: run {
        callSomeOtherFun()
        println("a is null case")
    }
}

fun callSomeOtherFun() = println("callSomeOtherFun")
fun nullable(): String? {
    return null
}
```
Even in the above example([playground link](https://pl.kotl.in/KcYyNLOKL)), all the print statements will be executed
2. `Or just with an Elvis operator is enough?` issue will reproduce with Elvis operator as well [playground link](https://pl.kotl.in/pRQHoNZ51). Here the issue is that most people associate `?.let { <expr1> } ?: <expr2>` as null check if else which is not correct
3. `This also seems to step a bit on unnecessaryLet.` If you are suggesting adding this logic to `unnecessaryLet` then that will make `unnecessaryLet` to other things than checking if `let` is unnecessary or not and overcomplicated that rule

**Comment by marschwar** [comment]
I don't think Brais was trying to be dismissive of your rule suggestion. I also think the way you described it it is too specific.  I think we all agree, that the code in question really should be flagged as hard to understand and error prone. The question remains: What is the pattern here that makes it "bad"?

```kotlin
a?.apply {
    println("a is not null case")
    nullable()
} ?: run {
    callSomeOtherFun()
    println("a is null case")
}
```

Does this have the same problem even if there is no `let`?  How about this:
```kotlin
a?.run {
    println("a is not null case")
    nullable()
} ?: callSomeOtherFun()
```

or this

```kotlin
a?.let {
    println("a is not null case")
    nullable()
}?.run {
    callSomeOtherFun()
    println("a is null case")
}
```

It is similar but also quite different. According to your initial post the one with `apply` would not be a problem as the `run`  statement does not depend on the outcome of `nullable()` (even though I would still consider this code hard to read).  But the last 2 examples would in fact be a violation.  Can you think of a clear set of rules to apply?

**Comment by BraisGabin** [comment]
Yes, that's exactly what I was asking for. I like the idea of the rule, but I would like to know exactly what are we looking for so we can formulate the rule correctly.

Maybe this rule is not about `let`. It is about any function that can return a `null`. Example:

```kotlin
foo?.maybeNullable() ?: doSomething()
```

No `let` involved but the same issue is present, I think. Should we look for this case too?

And this remembers me to other rule that I (usually) like: [`ReplaceSafeCallChainWithRun`](https://detekt.dev/docs/1.21.0/rules/complexity#replacesafecallchainwithrun). I say "usually" because there are some trivial cases where `ReplaceSafeCallChainWithRun` is a bit annoying. And probably the rule that we are talking about will have the same "problem". It could help a lot to reduce complexity but in some trivial cases it could be "annoying".

**Comment by atulgpt** [comment]
Hi @BraisGabin, @marschwar thanks for your inputs. Regarding `What is the pattern here that makes it "bad"?` I think the code snippets shared by you should also qualify for the rule. I think I listed down the most common pattern which bit me multiple times.

@BraisGabin regarding `foo?.maybeNullable() ?: doSomething()` I think this would be a little bit restrictive considering examples like `foo?.maybeNullable() ?: doSomethingIfAnyChainIsNull()`. While scoping operators should be fine to be caught while looking through the code we generally focus on scoping operators to understand code flow(ignoring the last statement of each scoping block). Or should we make this configurable?

**Comment by BraisGabin** [comment]
Then, what is the  difference between

```kotlin
foo?.maybeNullable() ?: doSomething()
```

and 

```
foo?.let { maybeNullable() } ?: doSomething
// or 
foo?.run { maybeNullable() } ?: doSomething
```

Are we only going to raise when it is a scoped function? Are we going to raise when the return type of the function is a generic? I mean: `let` returns `T` while `maybeNullable()` on the first example could return `String?`. One is a generic and the other is a known value.

These type of questions are the ones that I want to solve before this is implemented.

-----

My opinion: I wouldn't make this rule only for scope functions. The underlying problem happens on both cases. But I also agree that maybe that's a bit too restrictive. For that reason I will add something to make it configurable. But when it is "clearly wrong" and when it is "maybe wrong but acceptable". And how will we call that configuration so the users understand when the issue will be raised?

**Comment by atulgpt** [comment]
Hi, @BraisGabin sorry for the late reply. To answer your question of `Then, what is the difference between`, it just that generally, people consider `foo?.let { maybeNullable() } ?: doSomething` as an alternative to if else on condition that `foo` is null or not null. 

This consideration is not very common across other scope operators. With this rule, I wanted to tackle that issue.

**Comment by BraisGabin** [comment]
Now I understand your request! But I think that we should flag that even if we change `maybeNullable()` for `alwayNoNullable()`. You shouldn't use `let` as an `if/else`.

`let` is a PITA. It's so difficult to draw a line when a usage of `let` have sense and make the code better and when `let` just obfuscates the code.

`UnnecessaryLet` needs a reimplementation in my honest opinion. I tried it several times because it is my favorte rule on detekt. But I always end up in the same spot: where should I draw the line?

So, I would vote to close this issue and open an issue to talk about `UnnecessaryLet`. Create a list of valid and unvalid usages of `let`. And an explanation for each of them. And then we reimplement `UnnecessaryLet`.

**Comment by atulgpt** [comment]
I kind of agree with you @BraisGabin. But TBH `let` really feels very convenient tool to use as `if not null/else`. And it works well and makes the code readable except in the above case that I mentioned. Are you suggesting making `if not null/else` as invalid usage for let? I think that would make all users turn off the `UnnecessaryLet` rule. I think even Kotlin doc suggests([source](https://kotlinlang.org/docs/scope-functions.html#function-selection)) using `let` for non-nullable form of an object
<img width="769" alt="Screenshot 2024-02-25 at 4 53 21 PM" src="https://github.com/detekt/detekt/assets/11026203/850cf567-72ee-4c43-a7ed-0b3e39335625">
.

**Comment by BraisGabin** [comment]
It depends. Not all the null checks are the same. There are some that have sense. But as I said, draw the line is really difficult.

**Comment by atulgpt** [comment]
Not stale, as PR is open for review
